src/main/java/org/jboss/logmanager/handlers/WriterHandler.java

```diff
@@ -54,24 +67,46 @@ protected void doPublish(final ExtLogRecord record) {
             reportError("Formatting error", ex, ErrorManager.FORMAT_FAILURE);
             return;
         }
-        if (formatted.length() == 0) {
+        if (formatted.isEmpty()) {
             // nothing to write; don't bother
             return;
         }
         try {
+            waitingWritersHandle.getAndAdd(this, 1);
             lock.lock();
             try {
                 if (writer == null) {
+                    // nothing to write to
+                    waitingWritersHandle.getAndAdd(this, -1);
                     return;
                 }
-                preWrite(record);
+                try {
+                    preWrite(record);
+                } catch (Throwable t) {
+                    // the decrement was missed
+                    waitingWritersHandle.getAndAdd(this, -1);
+                    throw t;
+                }
+                // writer may have been changed by preWrite; re-check it
                 final Writer writer = this.writer;
                 if (writer == null) {
+                    // nothing to write to
+                    waitingWritersHandle.getAndAdd(this, -1);
                     return;
                 }
-                writer.write(formatted);
-                // only flush if something was written
-                super.doPublish(record);
+                try {
+                    writer.write(formatted);
+                } catch (Throwable t) {
+                    // the decrement was missed
+                    waitingWritersHandle.getAndAdd(this, -1);
+                    throw t;
+                }
+                // only flush if something was written and we're the last one
+                int remainingWaitersIncludingUs = (int) waitingWritersHandle.getAndAdd(this, -1);
+                // at this point the count is reconciled
+                if (remainingWaitersIncludingUs == 1 && isAutoFlush()) {
```

Comment:
Might be able to get rid of waitingWriters and the varhandle by just calling [lock.hasQueuedThreads()](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/locks/ReentrantLock.html#hasQueuedThreads()).  Since lock.lock() is called you should have to worry about cancellation that is mentioned in the API docs.